tests/governance_js.py

```diff
@@ -1205,3 +1207,145 @@ def make_script(table_name, kind):
+
+        first = consortium.get_any_active_member().propose(primary, proposal_body)
+        response = booster.vote(primary, first, ballot)
+        assert response.status_code == 200
+        response = turncoat.vote(primary, first, ballot)
+        assert response.status_code == 200
+
+        second = consortium.get_any_active_member().propose(primary, proposal_body)
+        response = booster.vote(primary, second, ballot)
+        assert response.status_code == 200
+        response = fairweather.vote(primary, second, ballot)
+        assert response.status_code == 200
+
+        third = consortium.get_any_active_member().propose(primary, proposal_body)
+        response = booster.vote(primary, third, ballot)
+        assert response.status_code == 200
+        # Votes against! Loses crony points!
+        response = turncoat.vote(
+            primary,
+            third,
+            json.dumps(
+                {
+                    "ballot": "export function vote (rawProposal, proposerId) { return false }"
+                }
+            ),
+        )
+        assert response.status_code == 200
+        response = fairweather.vote(primary, third, ballot)
+        assert response.status_code == 200
+
+        LOG.info("Confirm that finalVotes is present in submit-ballot response")
+        body = response.body.json()
+        assert "finalVotes" in body, body
+
+        LOG.info("Confirm that finalVotes is present in get-proposal response")
+        body = consortium.get_proposal_raw(primary, third.proposal_id)
+        assert "finalVotes" in body, body
+
+        LOG.info("Confirm that expected log lines were emitted")
+        expected_lines = [
+            f"Member {booster.service_id} has 3 points",
+            f"Member {fairweather.service_id} has 2 points",
+            f"Member {turncoat.service_id} has 0 points",
+        ]
+
+        for line in expected_lines:
+            assert primary.check_log_for_error_message(
+                line
+            ), f"Not found in stdout: {line}"
+
+    return network
+
+
+@reqs.description("Test final proposal description written to KV")
+def test_ledger_governance_invariants(network, args):
+    node = network.nodes[0]
+    ledger_dirs = node.remote.ledger_paths()
+
+    ledger = ccf.ledger.Ledger(ledger_dirs)
+
+    LOG.info("Completed proposals contain final_vote for each submitted ballot")
+    table_name = "public:ccf.gov.proposals_info"
+    for transaction in ledger.transactions():
+        public_tables = transaction.get_public_domain().get_tables()
+        if table_name not in public_tables:
+            continue
+
+        for _, raw_proposal in public_tables[table_name].items():
+            if raw_proposal is None:
+                # This is a deletion
+                continue
+
+            proposal = json.loads(raw_proposal)
+
+            state = proposal["state"]
+            if state in ("Open", "Withdrawn", "Dropped"):
```

Comment:
I've added this, but in doing so discovered that we actually _don't_ have an instance of `Failed`, because if we throw in `apply()` we deliberately don't write the result to the KV:

```
            if (proposal_info->state == ProposalState::FAILED)
            {
              // If the proposal failed to apply, we want to discard the tx and
              // not apply its side-effects to the KV state, because it may have
              // failed mid-execution (eg - thrown an exception), in which case
              // we do not want to apply partial writes
              detail::set_gov_error(
                ctx.rpc_ctx,
                HTTP_STATUS_INTERNAL_SERVER_ERROR,
                ccf::errors::InternalError,
                fmt::format("{}", proposal_info->failure));
              return;
            }
```

In other words, we can't extract "writing a Failed proposal" from any partial writes that `apply()` may have created before it threw. Maybe we should add a shim layer so that we _can_ make this distinction, maybe we should remove the `Failed` state now. Unclear to me which is preferable, but fixing is out-of-scope for this PR.